From e1adbd34286f3da104d4d2fae2c6ff3be294a813 Mon Sep 17 00:00:00 2001
From: Riccardo Rizzo <r.rizzo@arduino.cc>
Date: Tue, 7 Sep 2021 12:33:02 +0200
Subject: [PATCH 106/204] Add APIs to start and stop gnss functionality

the usage of start and stop should allows for switching the usage of gnss with respect for lte gsm operations
---
 .../GEMALTO_CINTERION_CellularStack.cpp       | 45 +++++++++++++++++--
 .../GEMALTO_CINTERION_CellularStack.h         |  6 ++-
 2 files changed, 46 insertions(+), 5 deletions(-)

diff --git a/connectivity/drivers/cellular/GEMALTO/CINTERION/GEMALTO_CINTERION_CellularStack.cpp b/connectivity/drivers/cellular/GEMALTO/CINTERION/GEMALTO_CINTERION_CellularStack.cpp
index 917accc916..6961e40f38 100644
--- a/connectivity/drivers/cellular/GEMALTO/CINTERION/GEMALTO_CINTERION_CellularStack.cpp
+++ b/connectivity/drivers/cellular/GEMALTO/CINTERION/GEMALTO_CINTERION_CellularStack.cpp
@@ -148,17 +148,54 @@ void GEMALTO_CINTERION_CellularStack::urc_gnss() {
     }
 }
 
-void GEMALTO_CINTERION_CellularStack::startGNSS(mbed::Callback<void(char*)> gnss_cb) {
+void GEMALTO_CINTERION_CellularStack::beginGNSS(mbed::Callback<void(char*)> gnss_cb) {
     _at.lock();
     _gnss_cb = gnss_cb;
     _at.at_cmd_discard("^SGPSC", "=", "%s%d", "Engine/StartMode", 0);
-    _at.at_cmd_discard("^SGPSC", "=", "%s%d", "Engine", 3);
-    //_at.at_cmd_discard("^SGPSC", "=", "%s%s", "Nmea/Output", "off");
-    _at.at_cmd_discard("^SGPSE", "=", "%s%s", "Nmea/Urc", "on");
+    _at.at_cmd_discard("^SGPSC", "=", "%s%d", "Engine", 0);
+    _at.at_cmd_discard("^SGPSC", "=", "%s%s", "Nmea/Urc", "off");
     _at.clear_error();
     _at.unlock();
 }
 
+void GEMALTO_CINTERION_CellularStack::endGNSS() {
+    _at.lock();
+    _gnss_cb = nullptr;
+    _at.clear_error();
+    _at.unlock();
+}
+
+void GEMALTO_CINTERION_CellularStack::startGNSS() {
+    _at.lock();
+    _engine = false;
+    _at.cmd_start_stop("^SGPSC", "=", "%s%d", "Engine", 3);
+    _at.resp_start("^SGPSC: \"Engine\",");
+
+    char respEng[2];
+    int resp_len = _at.read_string(respEng, sizeof(respEng));
+    if (strcmp(respEng, "3") != 0) {
+        _engine = false;
+        _at.at_cmd_discard("^SGPSC", "=", "%s%d", "Engine", 0);
+        _at.at_cmd_discard("^SGPSC", "=", "%s%s", "Nmea/Urc", "off");
+        return;
+    }
+    _engine = true;
+    _at.at_cmd_discard("^SGPSC", "=", "%s%s", "Nmea/Urc", "on");
+    _at.clear_error();
+    _at.unlock();
+}
+
+void GEMALTO_CINTERION_CellularStack::stopGNSS() {
+    if(_engine) {
+        _at.lock();
+        _gnss_cb = nullptr;
+        _at.at_cmd_discard("^SGPSC", "=", "%s%s", "Nmea/Urc", "off");
+        _at.at_cmd_discard("^SGPSC", "=", "%s%d", "Engine", 0);
+        _at.clear_error();
+        _at.unlock();
+        _engine = false;
+    }
+}
 
 nsapi_error_t GEMALTO_CINTERION_CellularStack::socket_stack_init()
 {
diff --git a/connectivity/drivers/cellular/GEMALTO/CINTERION/GEMALTO_CINTERION_CellularStack.h b/connectivity/drivers/cellular/GEMALTO/CINTERION/GEMALTO_CINTERION_CellularStack.h
index 163fa3193f..ffaed50184 100644
--- a/connectivity/drivers/cellular/GEMALTO/CINTERION/GEMALTO_CINTERION_CellularStack.h
+++ b/connectivity/drivers/cellular/GEMALTO/CINTERION/GEMALTO_CINTERION_CellularStack.h
@@ -36,7 +36,10 @@ public:
 
     void lock();
     void unlock();
-    void startGNSS(mbed::Callback<void(char*)> gnss_cb);
+    void beginGNSS(mbed::Callback<void(char*)> gnss_cb);
+    void endGNSS();
+    void startGNSS();
+    void stopGNSS();
 
 protected:
 
@@ -78,6 +81,7 @@ private:
     const char *_apn;
     const char *_user;
     const char *_password;
+    bool _engine;
 
     mbed::Callback<void(char*)> _gnss_cb;
 
-- 
2.39.1

